[Protoc] Show error on duplicate json names

YuryBandarchuk16 opened the pull request:
When having two fields with the same json name `protoc` did not fail with error neither showed a warning.

For example:
```proto
syntax = "proto3";
message Hi {
    int32 f1 = 1 [json_name = "fooBar"];
    int32 f2 = 2 [json_name = "fooBar"];
}
```

This MR shows warning when using `protoc` in the following format:
>> dup_json_test.proto: warning: Fields with names "Hi.f2" and "Hi.f1" have the same JSON name "fooBar". This may cause problems.

I decided to show warning instead of error in order to save backwards compatibility. It could be replaced with error later on.

cc: @xfxyjwf @acozzette @TeBoring @dsnet 


Fix #5063

dsnet commented on src/google/protobuf/descriptor.cc:
```diff
@@ -5894,6 +5894,21 @@ void DescriptorBuilder::ValidateProto3Message(
     }
   }
+
+  HASH_MAP<string, const FieldDescriptor*> reserved_json_name_map;
```
This problem affects proto2 as well, so I'm not sure the `ValidateProto3Message` function is the right place to add this.

dsnet commented:
No tests?

Sigh... I noted this issue a while ago. I would love to see this be an error and not a warning, but agree that this can break existing cases.

YuryBandarchuk16 commented:
@dsnet Thank you for your comments! I have updated the tests which got broken after switching from warning to error and added more tests. Will appreciate if you take a look again!

wangkirin commented:
Hi all, I found this problem still exists in the master until now,  and your commit also conflicts with the latest. I would love to finish fixing this problem.

TeBoring commented on src/google/protobuf/descriptor.cc:
```diff
@@ -4576,6 +4576,21 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
     }
   }
 
+  HASH_MAP<string, const FieldDescriptor*> reserved_json_name_map;
```
Only proto3 message should have this check

TeBoring commented:
Feel free to reopen after sync the change.
